### ISSUE TITLE
"The forked VM terminated without properly saying goodbye. VM crash or System.exit called?"

### ISSUE THREAD

**Issue by ams-tschoening** [wicketstuff/core#676]
I'm trying to build Wicket Stuff 8.5.0 and got the following error during running the tests:

    [ERROR] org.apache.maven.surefire.booter.SurefireBooterForkException: The forked VM terminated without properly saying goodbye. VM crash or System.exit called?
    [ERROR] Command was cmd.exe /X /C ""C:\Program Files\Java\jdk-8\bin\java.exe" -Xmx1024m -jar C:\Users\tschoening\AppData\Local\Temp\surefire899818338470464788\surefirebooter728873680916515324.jar C:\Users\tschoening\AppData\Local\Temp\surefire899818338470464788 2019-09-30T11-28-48_299-jvmRun1 surefire815810429636484508tmp surefire_461001889288217241939tmp"
    [ERROR] Process Exit Code: 0
    [ERROR] Crashed tests:
    [ERROR] org.wicketstuff.jamon.request.cycle.JamonMonitoredRequestCycleTest

The error always happens exactly the same way, especially regarding the process exit status, and happens in the current master as well. Because the exit status is always the same `0` and increasing e.g. `-Xmx1024m` to `-Xmx4096m` doesn't change a thing, I don't think that the VM is really crashing, but instead someone is calling `System.exit` somewhere. Couldn't find such statement in the project source, though, only in `jamon-examples`, but that doesn't seem to be involved.

Debugging this further revealed that ignoring the test [shouldNotMonitorJamonAdminPageItSelf](https://github.com/wicketstuff/core/blob/master/jamon-parent/jamon/src/test/java/org/wicketstuff/jamon/request/cycle/JamonMonitoredRequestCycleTest.java#L72) makes the build succeed:

         @Ignore// broken in Wicket 8.0. Needs debugging!
         @Test
         public void shouldNotMonitorJamonAdminPageItSelf()
         {
                 wicketTester.startPage(JamonAdminPage.class);
                 assertEquals(0, MonitorFactory.getMonitor("JamonAdminPage", "ms.").getHits(), 0);
                 FormTester formTester = wicketTester.newFormTester("adminPanel:adminForm");
                 formTester.setValue("monitorLabel", "J");
                 wicketTester.executeAjaxEvent("adminPanel:adminForm:monitorLabel", "keyup");
                 assertEquals(0, MonitorFactory.getMonitor("JamonAdminPage", "ms.").getHits(), 0);
         }

That's of interest because that file already contains some other ignored test. But if I enable those again, they result in different errors, so are most likely not related to this problem. Additional debugging revealed that changing `pom.xml` to include the following option for surefire makes the build succeed as well, even WITHOUT ignoring the former test anymore:

    <reuseForks>false</reuseForks>

Switching between `true` and `false` make the problem occur/disappear. That's interesting, because directly after the build fails running the command mentioned in the error message in another shell leads to the following output:

    cmd.exe /X /C ""C:\Program Files\Java\jdk-8\bin\java.exe" -Xmx1024m -jar C:\Users\tschoening\AppData\Local\Temp\surefire3649246175279820451\surefirebooter4577796592449395223.jar C:\Users\tschoening\AppData\Local\Temp\surefire3649246175279820451 2019-09-30T16-38-42_368-jvmRun1 surefire9181599006974311426tmp surefire_07065828248506771667tmp""

    3,1,windows-1252,[org.apache.wicket.page.PageAccessSynchronizer] DEBUG - 'main' released lock to page with id '1'\0D\0A
    3,1,windows-1252,[org.apache.wicket.page.PageAccessSynchronizer] DEBUG - 'main' notifying blocked threads\0D\0A
    3,1,windows-1252,[org.apache.wicket.protocol.http.servlet.ServletWebRequest] DEBUG - Calculating context relative path from: context path '/context'\2C filterPrefix 'servlet/'\2C uri '/context/servlet/'\0D\0A
    6,1,org.wicketstuff.jamon.request.cycle.JamonMonitoredRequestCycleTest,shouldCreateTwoMonitorsForPagesThatAreNavigatedToTwice(org.wicketstuff.jamon.request.cycle.JamonMonitoredRequestCycleTest),null,null,null
    2,1,org.apache.maven.surefire.junit4.JUnit4Provider,org.wicketstuff.jamon.request.cycle.JamonMonitoredRequestCycleTest,null,null,null
    Z,0,BYE!
    
Which very much reads like a proper "goodbye" like mentioned in the error message at the beginning. In the end, either of both works, either additionally ignoring the mentioned test or changing the fork behaviour of surefire. My feeling is that for some reason surefire is not able to properly recognize that goodbye in some combinations of tests anymore if those were run within the same JVM.

My environment:

* Windows 10 Pro x86-64 1903 Build 18362.356
* Maven 3.6.2
* openjdk version "1.8.0_222"
  * OpenJDK Runtime Environment (AdoptOpenJDK)(build 1.8.0_222-b10)
  * OpenJDK 64-Bit Server VM (AdoptOpenJDK)(build 25.222-b10, mixed mode)

I've reported this problem on the [wicket-dev-list](https://lists.apache.org/thread.html/e3ab0595fe0ddf35fa374349ec61867fd9ce1a2983ff78239ed1436a@%3Cdev.wicket.apache.org%3E) as well.

[WicketStuff 8.5.0 Jamon VM-crash.txt](https://github.com/wicketstuff/core/files/3672208/WicketStuff.8.5.0.Jamon.VM-crash.txt)

**Comment by ams-tschoening** [comment]
<reuseForks>false</reuseForks>

That doesn't work in the `pom.xml` in the root, because tests for Quickview fail in this case. Moving that line into Jamon's `pom.xml` only seems to work.

**Comment by ams-tschoening** [comment]
After switching to 8.7.0-SNAPSHOT I'm again running into this problem, only weirder: Switching the reuse-option from `false` to `true` makes the test always fail in my environment simply by issuing `mvn test -pl :wicketstuff-jamon`. But having it set to `false` doesn't guarantee a succeeding build anymore, but instead it seems that calling `mvn package` fails always. `mvn test` instead seems to succeed always as well as `mvn package -X`. Yes, adding `-X` makes a difference for me. Additionally I need to build everything to trigger the problem with reuse-option `false`, so far I was unable to trigger it with only building some of the projects. Tried with 10 before `jamon` and all afterwards, but things succeeded. Doing the same without restrictions to the projects and the build failed again.

So I guess `reuseForks` only worked around the real problem of the test under some circumstances. The interesting thing is that the tests themself don't seem to fail, but instead JVM-management seems to be the real problem for some reason. Surefire simply doesn't get the expected answer after the tests succeed, but the JVM really doesn't crash. There is something else going on...

Surefire seems to execute the following command line in the end:

    cmd.exe /X /C ""C:\Program Files\Java\jdk-8\bin\java.exe" -Xmx1024m -jar C:\Users\tschoening\AppData\Local\Temp\surefire6535499946100577518\surefirebooter2200936470851871489.jar C:\Users\tschoening\AppData\Local\Temp\surefire6535499946100577518 2019-10-03T09-50-14_555-jvmRun1 surefire6314078221206684409tmp surefire_07911726589157278885tmp""

And as said, there's no test error or anything like that in the output. The last lines are the following:

> 3,1,windows-1252,[org.apache.wicket.page.PageAccessSynchronizer] DEBUG - 'main' released lock to page with id '1'\0D\0A
> 3,1,windows-1252,[org.apache.wicket.page.PageAccessSynchronizer] DEBUG - 'main' notifying blocked threads\0D\0A
> 3,1,windows-1252,[org.apache.wicket.protocol.http.servlet.ServletWebRequest] DEBUG - Calculating context relative path from: context path '/context'\2C filterPrefix 'servlet/'\2C uri '/context/servlet/'\0D\0A
> 6,1,org.wicketstuff.jamon.request.cycle.JamonMonitoredRequestCycleTest,shouldCreateTwoMonitorsForPagesThatAreNavigatedToTwice(org.wicketstuff.jamon.request.cycle.JamonMonitoredRequestCycleTest),null,null,null
> 2,1,org.apache.maven.surefire.junit4.JUnit4Provider,org.wicketstuff.jamon.request.cycle.JamonMonitoredRequestCycleTest,null,null,null
> Z,0,BYE!
> 

What's strange is that after the last newline has been printed, the JVM seems to stay open for quite some time.

[surefirebooter2200936470851871489.jar.zip](https://github.com/wicketstuff/core/files/3685020/surefirebooter2200936470851871489.jar.zip)
[surefire_07911726589157278885tmp.txt](https://github.com/wicketstuff/core/files/3685021/surefire_07911726589157278885tmp.txt)
[surefire6314078221206684409tmp.txt](https://github.com/wicketstuff/core/files/3685022/surefire6314078221206684409tmp.txt)

**Comment by ams-tschoening** [comment]
Guten Tag Thorsten Schöning,
am Donnerstag, 3. Oktober 2019 um 09:45 schrieben Sie:

> There seems to be some weird sideeffect somewhere making the test
> fail in the end.[...]

I hopefully found it, the runtime of the test itself. I'm always
invoking the same command:

> mvn test -pl :wicketstuff-jamon

That always fails with the default options of reusing forks:

> <reuseForks>true</reuseForks>

It at least mostly succeeds when that option is set to false. But
that's only an unreliable workaround for the real problem, I can
switch between failing/succeeding build of jamon by using the
following option only as well:

> <forkedProcessExitTimeoutInSeconds>2</forkedProcessExitTimeoutInSeconds>
vs.
> <forkedProcessExitTimeoutInSeconds>200</forkedProcessExitTimeoutInSeconds>

I've removed "reuseForks" and only used that option above with those
two different values and can reliably switch between a failing and
succeeding build. More important, the test really runs longer, I get
far more output on STDOUT and stuff.

The default value of that option is 30 and with default settings the
test always fails after ~35 to ~40 seconds. With "reuseForks" set to
false the tests succeed mostly, but their runtime is only about 50 -
60 seconds as well. With setting "forkedProcessExitTimeoutInSeconds"
much higher the tests run ~80 to ~90 seconds. Which is a lot longer
than 30, but far less than 200 and I don't have any additional waiting
time or else.

> [WARNING] Tests run: 7, Failures: 0, Errors: 0, Skipped: 2, Time elapsed: 33.075 s - in org.wicketstuff.jamon.request.cycle.JamonMonitoredRequestCycleTest
> [INFO]
> [INFO] Results:
> [INFO]
> [WARNING] Tests run: 24, Failures: 0, Errors: 0, Skipped: 2
> [INFO]
> [INFO] ------------------------------------------------------------------------
> [INFO] BUILD SUCCESS
> [INFO] ------------------------------------------------------------------------
> [INFO] Total time:  01:24 min
> [INFO] Finished at: 2019-10-03T10:39:31+02:00
> [INFO] ------------------------------------------------------------------------

How long does this test run for you?

The option I'm using is documented like the following:

> After the test-set has completed, the process executes
> java.lang.System.exit(0) which starts shutdown hooks. At this point
> the process may run next 30 seconds until all non daemon Threads
> die. After the period of time has elapsed, the process kills itself
> by java.lang.Runtime.halt(0). The timeout of 30 seconds can be
> customized by configuration parameter
> forkedProcessExitTimeoutInSeconds.

https://maven.apache.org/surefire/maven-surefire-plugin/examples/shutdown.html

That makes somewhat sense to what I see how the test behaves. But that
would mean that tests are still running AFTER Surefire has come to the
onclusion they have finished already?

Mit freundlichen Grüßen,

Thorsten Schöning

-- 
Thorsten Schöning       E-Mail: Thorsten.Schoening@AM-SoFT.de
AM-SoFT IT-Systeme      http://www.AM-SoFT.de/

Telefon...........05151-  9468- 55
Fax...............05151-  9468- 88
Mobil..............0178-8 9468- 04

AM-SoFT GmbH IT-Systeme, Brandenburger Str. 7c, 31789 Hameln
AG Hannover HRB 207 694 - Geschäftsführer: Andreas Muchow

**Comment by ams-tschoening** [comment]
Guten Tag Thorsten Schöning,
am Donnerstag, 3. Oktober 2019 um 10:44 schrieben Sie:

> How long does this test run for you?

I made some additional tests and was easily able to trigger the
problem on Windows 7 and Windows 10 1809 using Maven 3.6.2 and Java
1.8.0_222.

I did some additional testing on UB 16.04 using Maven 3.3.9 and the
same Java and was able to trigger the problem there as well under the
following conditions: "forkedProcessExitTimeoutInSeconds" was set to
10 seconds and the test was executed using the GUI of VMware
Workstation. The latter part is important, because the same exact
setting, but executing the test using SSH succeeds. Redirecting STDOUT
of running the test using the VMware-GUI succeeds as well. 

So I tried the same on Windows: Redicrecting STDOUT to some file makes
the test succeed as well on all formerly failing Windows. Using e.g. GIT
Bash with its own shell to run the test makes it succeed as well.

This makes two things clear in my opinion: There are background
threads creating a lot of output on STDOUT. That's why "forked*"
makes a difference at all, because either those background threads
have enough time to properly finish or not.

But the root cause of the problem seems to be reading that STDOUT by
the surrounding shell, which is either fast enough or not. The default
shells on Linux seem to be much faster in doing so compared to cmd.exe
and Powershell on Windows. The long running tests I had seen with my
additional Windows-tests were because of using RDP and VMware-GUI:
Printing STDOUT in those contexts is surely synchronized at some
point, therefore pretty slow and the tests need to run far longer for
the background-threads to properly finish. That's why things succeed
if STDOUT gets redirected on Windows, if GIT Bash is used and why the
same problem can be triggered on Linux if one is able to make the
surrounding shell reading STDOUT slower somehow.

Mit freundlichen Grüßen,

Thorsten Schöning

-- 
Thorsten Schöning       E-Mail: Thorsten.Schoening@AM-SoFT.de
AM-SoFT IT-Systeme      http://www.AM-SoFT.de/

Telefon...........05151-  9468- 55
Fax...............05151-  9468- 88
Mobil..............0178-8 9468- 04

AM-SoFT GmbH IT-Systeme, Brandenburger Str. 7c, 31789 Hameln
AG Hannover HRB 207 694 - Geschäftsführer: Andreas Muchow

**Comment by martin-g** [comment]
Please try with latest `wicket-8.x`. I've reduced the amount of writes to stdout by logging only WARNs and ERRORs

**Comment by ams-tschoening** [comment]
Logging less fixes the problem as well. The problem I see with this approach is that if someone really wants to debug the test because of some problems and logs more, things break again. Additionally raising the timeout seems to be an always working solution, so I suggest doing both with a reasonable high value of `500`.

Depending on where background threads are used during tests as well, it might even be a good idea to change the timeout in the main `pom.xml` already and link to this issue. That main file already contains additional settings, where this one could easily be added:

	<testFailureIgnore>false</testFailureIgnore>
	<argLine>-Xmx1024m</argLine>

**Comment by martin-g** [comment]
So far you are the only one having problems with this test (in several years).
If you don't have time and desire to debug more why Windows is not happy with more log output I'd prefer to close the issue without adding more JVM/JUnit/Surefire options which may affect other modules.

**Comment by ams-tschoening** [comment]
> So far you are the only one having problems with this test (in several years).

Like I seemed to be the only one using paths with spaces triggering a bug in your code, so that doesn't tell much.

> If you don't have time and desire to debug more why Windows is not happy with more log output[...]

You have noticed that I was able to trigger the exact same problem on UB 16.04 as well? I'm pretty sure to already have found the root cause of the problem, that's why I copied my mail:

You are using background threads somewhere not synced to test execution, so the tests itself finish before the background threads got rid of all their STDOUT. Surefire allows those threads to still run some amount of time and in some environments this is enough and in others it isn't, simply depending on how fast STDOUT is read. That perfectly fits to what I've debugged and is even documented by Surefire:

> After the test-set has completed, the process executes java.lang.System.exit(0) which starts shutdown hooks. At this point the process may run next 30 seconds until all non daemon Threads
die.

https://maven.apache.org/surefire/maven-surefire-plugin/examples/shutdown.html

You need to assure that your background threads are properly finished as well before ending each individual test and this simply isn't the case currently. The second best approach is to increase the time Surefire gives them to finish on their own.

**Comment by martin-g** [comment]
We are all volunteers here!
Please do not associate neither Wicketstuff projects nor Wicket itself with **me**. `You are using background threads ...` && `You need to assure that your background threads are properly finished...`. I, personally, have no background threads. WicketStuff Jamon module may have such threads but I **don't**. I do not use neither Jamon nor WicketStuff Jamon. Someone from the community decided to use Jamon and integrate it Wicket and to donate the code so others can benefit from it.
If no one have noticed that the software does not work properly on Windows (or any other OS) is most probably because: 1) no one ever used it on Windows; 2) no one actually uses it at all.

I have provided you with a stable solution, so you can build the project without extra configuration changes. If this workaround is not enough for you then feel free to continue investigating! Since I cannot reproduce the problem I cannot help you more!

**Comment by ams-tschoening** [comment]
> Please do not associate neither Wicketstuff projects nor Wicket itself with me.

I was using `you` simply because you used it as well and was referring to one of the gatekeepers of this project this way. Instead of feeling offended, you might want to look at how your own response read after I spent days debugging this problem.

> I have provided you with a stable solution, so you can build the project without extra configuration changes.

But as said, it's not stable, it breaks instantly when I really need logging in the tests. If you insist to only reduce the log level, than you should at least revert the former PR introducing `reuseForks` entirely. Setting it to `true` is misleading, because that whole part in `pom.xml` is about `true` not working properly.

Merging my newer PR instead would be the better option of course: It keeps the already introduced XML, links to this issue in case others are affected as well and introduces an additional workaround for the problem even with debug-logging enabled.

**Comment by i-make-robots** [comment]
Hi!  I'm getting a maven surefire timeout bug in another project and this issue is as close as I could find to related issue.  (Thanks, google.)  Anyone know why this is happening or what can be done about it?  

I have isolated it down to a call to java.util.prefs.Preferences.get() which seems to hang indefinitely.  Oddly the surefire runs fine on my local windows box.  It's only in the Github Action for continuous integration that it hangs.  I am running surefire 3.0.0-M5 and I've tried different forkCount, reuseForks=false, useSystemClassLoader=true.... fwiw my issue ticket is https://github.com/MarginallyClever/Makelangelo-software/issues/363

I'm out of ideas and grasping at straws.  Thank you!
